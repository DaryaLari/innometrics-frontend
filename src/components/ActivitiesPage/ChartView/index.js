import React from 'react'
import Chart from 'react-google-charts'
import { connect } from 'react-redux'
import { activitiesSummarized } from '../../../helpers/selectors'
import Spinner from '../../Spinner'
import styles from './style.css'

class ChartView extends React.Component {
  render() {
    return (
      <div className={styles.chartView}>
        <Chart
          // width={'500px'}
          height={`${this.props.activities.length * 30}px`}
          chartType='BarChart'
          loader={<Spinner/>}
          data={[['Activity', 'Duration (min)', { role: 'annotation' }]]
            .concat(this.props.activities.map(a => {
              let time = Number(a.duration)
              const fromDays = 24 * 60 * 60,
                    fromHours = 60 * 60,
                    fromMinutes = 60
              const days = Math.floor(time / fromDays)
              time -= days * fromDays
              const hours = Math.floor(time / fromHours)
              time -= hours * fromHours
              const minutes = Math.floor(time / fromMinutes)
              const seconds = Math.round(time - minutes * fromMinutes)
              const formatted = (x) => (x < 10 ? '0' : '') + x
              return [
                a.executable_name, a.duration / 60,
                `${days > 0 ? days + 'days\n' : ''} ${formatted(hours)}:${formatted(minutes)}:${formatted(seconds)}`
              ]
            }))
          }
          options={{
            title: 'Time spend for activities',
            // chartArea: { width: '50%' },
            hAxis: {
              title: 'Time (min)',
            },
            vAxis: {
              title: 'Activity',
              minValue: 0,
            },
            legend: { position: 'none' },
            // bar: { groupWidth: '30px' },
          }}
        />
      </div>
    )
  }
}

const ConnectedChartView = connect(
  (state) => ({
    activities: activitiesSummarized(state)
  })
)(ChartView)

export default ConnectedChartView