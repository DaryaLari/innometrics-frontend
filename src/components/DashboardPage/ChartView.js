import React from "react";
import styles from "./style.css";
import { Bar, BarChart, CartesianGrid, ResponsiveContainer, Tooltip, XAxis, YAxis } from 'recharts'

class ChartView extends React.Component {
  state = {
    activities: this.props.activities.map(a => {
      return {
        actName: a.executable_name,
        duration: Date.parse(a.start_time) - Date.parse(a.end_time)
      }})
  }
  render() {
    return (
      <div className={styles.chartView}>
        <ResponsiveContainer width={'100%'} aspect={2}>
          <BarChart data={this.state.activities}
                         margin={{ top: 25, right: 50, left: 50, bottom: 25 }}>
          <CartesianGrid strokeDasharray="3 3" />
          <XAxis dataKey="actName" label={{ offset: 0, value: "Activity", position: 'bottom' }} height={10}  axisLine={false} tickLine={false} tick={false} />
          <YAxis label={{ offset: 13, value: "Duration (sec)", position: 'top' }} dataKey="duration" />
          <Tooltip payload={this.state.activities} />
          <Bar dataKey="duration" barSize={30} fill="#8884d8" />
        </BarChart>
        </ResponsiveContainer>
      </div>
    )
  }
}

export default ChartView;