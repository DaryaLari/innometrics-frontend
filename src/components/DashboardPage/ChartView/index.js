import React from 'react'
import { connect } from 'react-redux'
import { Bar, BarChart, CartesianGrid, ResponsiveContainer, Tooltip, XAxis, YAxis } from 'recharts'
import { activitiesSummarized } from '../../../helpers/selectors'
import styles from './style.css'

class ChartView extends React.Component {
  render() {
    return (
      <div className={styles.chartView}>
        <ResponsiveContainer width={'100%'} aspect={2}>
          <BarChart data={this.props.activities}
                    margin={{ top: 25, right: 50, left: 50, bottom: 25 }}>
            <CartesianGrid strokeDasharray='3 3' />
            <XAxis dataKey='executable_name'
                   label={{ offset: 0, value: 'Activity', position: 'bottom' }}
                   height={25}
                   axisLine={false}
                   tickLine={false}
                   interval={0}
                   tick={<CustomTick amount={this.props.activities.length}/>}
            />
            <YAxis label={{ offset: 13, value: 'Duration (sec)', position: 'top' }} dataKey='duration' />
            <Tooltip payload={this.props.activities} />
            <Bar dataKey='duration'
                 barSize={30}
                 fill='#8884d8'
                 onClick={this.props.onBarClick}
            />
        </BarChart>
        </ResponsiveContainer>
      </div>
    )
  }
}

class CustomTick extends React.Component{
  render(){
    const {x, y, payload, ...rest} = this.props
    let width = rest.width/rest.amount - 5
    return (
      <foreignObject className={styles.truncated}
                     x={x-width/2} y={y-10}
                     width={width} height={20}
      >
        <span title={payload.value}>{payload.value}</span>
      </foreignObject>

    )
  }
}

const ConnectedChartView = connect(
  (state) => ({
    activities: activitiesSummarized(state)
  })
)(ChartView)

export default ConnectedChartView