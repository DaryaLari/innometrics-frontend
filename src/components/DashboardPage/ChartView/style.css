.chartView {
    width: 50%;
}

.truncated {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
}