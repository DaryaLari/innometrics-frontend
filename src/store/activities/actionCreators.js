import {TYPES as ACTIVITIES_TYPES} from './actionTypes'
import {getRequest} from '../../helpers/api'

export const getActivitiesRequest = () => (dispatch, getState) => {
  dispatch({type: ACTIVITIES_TYPES.GET_ACTIVITIES_REQUEST})

  getRequest('/activity', {offset: 0, amount_to_return: 100}, true)
      .then((result) => {
        dispatch({
          type: ACTIVITIES_TYPES.GET_ACTIVITIES_SUCCESS,
          payload: {
            activities: result.data.activities,
            offset: 0,
            amount_to_return: 100
          }
        })
      })
      .catch((error) => {
        dispatch({type: ACTIVITIES_TYPES.GET_ACTIVITIES_FAILURE, error: error.data.message})
      })
}