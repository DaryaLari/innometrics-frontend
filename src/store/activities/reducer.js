import {TYPES as ACTIVITIES_TYPES} from './actionTypes'
import {TYPES as USER_TYPES} from '../user/actionTypes'

const initialState = {
  activities: [],
  offset: 0,
  amount_to_return: 0,

  activeRequest: false,
  failed: false,
  error: null
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ACTIVITIES_TYPES.INIT_ACTIVITIES:
    case USER_TYPES.LOGIN_SUCCESS:
    case USER_TYPES.LOGOUT_SUCCESS:
    case USER_TYPES.LOGOUT_FAILURE:
      return initialState


    case ACTIVITIES_TYPES.GET_ACTIVITIES_REQUEST:
      return {
        ...state,
        activeRequest: true,
        failed: false,
        error: null
      }

    case ACTIVITIES_TYPES.GET_ACTIVITIES_SUCCESS:
      return {
        activities: action.payload.activities,
        offset: action.payload.offset,
        amount_to_return: action.payload.amount_to_return,
        activeRequest: false,
        failed: false,
        error: null
      }

    case ACTIVITIES_TYPES.GET_ACTIVITIES_FAILURE:
      return {
        ...state,
        activeRequest: false,
        failed: true,
        error: action.error
      }

    default:
      return state
  }
}