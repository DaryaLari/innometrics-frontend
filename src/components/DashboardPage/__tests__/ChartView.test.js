import React from 'react'
import { shallow } from 'enzyme/build/index'
import {activities} from '../../../../___mocks___/activities'
import ChartView from '../ChartView'
import createStore from 'redux-mock-store'
import { Provider } from 'react-redux'

describe('Test ChartView component', () => {

  it('Test shallow render ChartView component', () => {

    const mockStore = createStore([]);
    const store = mockStore({})

    const wrapper = shallow(
      <Provider store={store}>
        <ChartView activities={activities} />
      </Provider>
    )

    expect(wrapper).toMatchSnapshot()
  })

})