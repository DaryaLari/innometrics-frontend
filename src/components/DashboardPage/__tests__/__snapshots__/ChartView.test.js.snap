// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ChartView component Test shallow render ChartView component 1`] = `
<Connect(ChartView)
  activities={
    Array [
      Object {
        "_id": "5bd1aa255f627d1a8324f9d0",
        "browser_title": "",
        "browser_url": "",
        "end_time": "2018-10-25 11:28:59",
        "executable_name": "file:///Users/<username>/Library/Developer/Xcode/DerivedData/InnometricsTransfer-ftpzsieukjkaktfvhcxnjxjwyjwi/Build/Products/Debug/InnometricsTransfer.app/Contents/MacOS/InnometricsTransfer",
        "ip_address": "192.168.1.253",
        "mac_address": "34:36:3b:c4:85:70",
        "start_time": "2018-10-25 11:29:33",
        "user": "5bbe21f95f627d8c7a5754ac",
      },
      Object {
        "_id": "5bd1aa255f627d1a8324f9c8",
        "browser_title": "",
        "browser_url": "",
        "end_time": "2018-10-25 11:33:47",
        "executable_name": "file:///Users/<username>/Library/Developer/Xcode/DerivedData/InnometricsTransfer-ftpzsieukjkaktfvhcxnjxjwyjwi/Build/Products/Debug/InnometricsTransfer.app/Contents/MacOS/InnometricsTransfer",
        "ip_address": "192.168.1.253",
        "mac_address": "34:36:3b:c4:85:70",
        "start_time": "2018-10-25 11:33:50",
        "user": "5bbe21f95f627d8c7a5754ac",
      },
      Object {
        "_id": "5bd1aa255f627d1a8324f9c7",
        "browser_title": "",
        "browser_url": "",
        "end_time": "2018-10-25 11:33:50",
        "executable_name": "file:///Applications/Xcode.app/Contents/MacOS/Xcode",
        "ip_address": "192.168.1.253",
        "mac_address": "34:36:3b:c4:85:70",
        "start_time": "2018-10-25 11:35:11",
        "user": "5bbe21f95f627d8c7a5754ac",
      },
      Object {
        "_id": "5bd1aa255f627d1a8324f9ca",
        "browser_title": "",
        "browser_url": "",
        "end_time": "2018-10-25 11:33:40",
        "executable_name": "file:///Users/<username>/Library/Developer/Xcode/DerivedData/InnometricsTransfer-ftpzsieukjkaktfvhcxnjxjwyjwi/Build/Products/Debug/InnometricsTransfer.app/Contents/MacOS/InnometricsTransfer",
        "ip_address": "192.168.1.253",
        "mac_address": "34:36:3b:c4:85:70",
        "start_time": "2018-10-25 11:33:45",
        "user": "5bbe21f95f627d8c7a5754ac",
      },
      Object {
        "_id": "5bd1aa255f627d1a8324f9cd",
        "browser_title": "",
        "browser_url": "",
        "end_time": "2018-10-25 11:29:54",
        "executable_name": "file:///Applications/Safari.app/Contents/MacOS/Safari",
        "ip_address": "192.168.1.253",
        "mac_address": "34:36:3b:c4:85:70",
        "start_time": "2018-10-25 11:29:56",
        "user": "5bbe21f95f627d8c7a5754ac",
      },
      Object {
        "_id": "5bd1aa255f627d1a8324f9ce",
        "browser_title": "",
        "browser_url": "",
        "end_time": "2018-10-25 11:29:45",
        "executable_name": "file:///Users/<username>/Library/Developer/Xcode/DerivedData/InnometricsTransfer-ftpzsieukjkaktfvhcxnjxjwyjwi/Build/Products/Debug/InnometricsTransfer.app/Contents/MacOS/InnometricsTransfer",
        "ip_address": "192.168.1.253",
        "mac_address": "34:36:3b:c4:85:70",
        "start_time": "2018-10-25 11:29:54",
        "user": "5bbe21f95f627d8c7a5754ac",
      },
      Object {
        "_id": "5bd1aa255f627d1a8324f9cf",
        "browser_title": "",
        "browser_url": "",
        "end_time": "2018-10-25 11:29:33",
        "executable_name": "file:///Applications/Safari.app/Contents/MacOS/Safari",
        "ip_address": "192.168.1.253",
        "mac_address": "34:36:3b:c4:85:70",
        "start_time": "2018-10-25 11:29:45",
        "user": "5bbe21f95f627d8c7a5754ac",
      },
      Object {
        "_id": "5bd1aa255f627d1a8324f9cc",
        "browser_title": "",
        "browser_url": "",
        "end_time": "2018-10-25 11:29:56",
        "executable_name": "file:///Applications/PyCharm.app/Contents/MacOS/pycharm",
        "ip_address": "192.168.1.253",
        "mac_address": "34:36:3b:c4:85:70",
        "start_time": "2018-10-25 11:30:00",
        "user": "5bbe21f95f627d8c7a5754ac",
      },
      Object {
        "_id": "5bd1aa255f627d1a8324f9cb",
        "browser_title": "",
        "browser_url": "",
        "end_time": "2018-10-25 11:30:00",
        "executable_name": "file:///Applications/Telegram.app/Contents/MacOS/Telegram",
        "ip_address": "192.168.1.253",
        "mac_address": "34:36:3b:c4:85:70",
        "start_time": "2018-10-25 11:33:40",
        "user": "5bbe21f95f627d8c7a5754ac",
      },
      Object {
        "_id": "5bd1aa255f627d1a8324f9c9",
        "browser_title": "",
        "browser_url": "",
        "end_time": "2018-10-25 11:33:45",
        "executable_name": "file:///Applications/Xcode.app/Contents/MacOS/Xcode",
        "ip_address": "192.168.1.253",
        "mac_address": "34:36:3b:c4:85:70",
        "start_time": "2018-10-25 11:33:47",
        "user": "5bbe21f95f627d8c7a5754ac",
      },
    ]
  }
/>
`;
